File: lib/ios/Models/AdaptyProduct.swift
//
//  AdaptyProduct.swift
//  RNAdapty
//
//  Created by Ivan on 12.07.2021.
//

import Adapty
import Foundation

struct AdaptyProductIos: Encodable {
  let discounts: [AdaptyProductDiscount]
  let isFamilyShareable: Bool
  let promotionalOfferId: String?
  let subscriptionGroupIdentifier: String?
  let localizedSubscriptionPeriod: String?
  let regionCode: String?
  let promotionalOfferEligibility: Bool
}

struct AdaptyProduct: Encodable {
  let vendorProductId: String
  let introductoryOfferEligibility: Bool
  let paywallABTestName: String?
  let paywallName: String?
  let localizedDescription: String
  let localizedTitle: String
  let price: Decimal
  let currencyCode: String?
  let currencySymbol: String?
  let regionCode: String?
  let subscriptionPeriod: AdaptySubscriptionPeriod?
  let introductoryDiscount: AdaptyProductDiscount?
  let localizedPrice: String?

  let ios: AdaptyProductIos

  init(_ product: ProductModel) {
    vendorProductId = product.vendorProductId
    introductoryOfferEligibility = product.introductoryOfferEligibility
    paywallABTestName = product.paywallABTestName
    paywallName = product.paywallName
    localizedDescription = product.localizedDescription
    localizedTitle = product.localizedTitle
    price = product.price
    currencyCode = product.currencyCode
    currencySymbol = product.currencySymbol
    regionCode = product.regionCode
    localizedPrice = product.localizedPrice

    if let productSubscriptionPeriod = product.subscriptionPeriod {
      subscriptionPeriod = AdaptySubscriptionPeriod.init(productSubscriptionPeriod)
    } else {
      subscriptionPeriod = nil
    }
    if let productIntroDiscount = product.introductoryDiscount {
      introductoryDiscount = AdaptyProductDiscount.init(productIntroDiscount)
    } else {
      introductoryDiscount = nil
    }

    ios = AdaptyProductIos(
      discounts: product.discounts.map {AdaptyProductDiscount.init($0) },
                           isFamilyShareable: product.isFamilyShareable,
                           promotionalOfferId: product.promotionalOfferId,
                           subscriptionGroupIdentifier: product.subscriptionGroupIdentifier,
                           localizedSubscriptionPeriod: product.localizedSubscriptionPeriod,
                           regionCode: product.regionCode,
                           promotionalOfferEligibility: product.promotionalOfferEligibility
                           )
  }
}

File: lib/src/coders/adapty-product.ts
import type { AdaptyProduct } from '../types';
import { Coder } from './coder';
import { AdaptyProductSubscriptionPeriodCoder } from './adapty-product-subscription-period';
import { AdaptyProductDiscountCoder } from './adapty-product-discount';
import { Properties } from './types';
import { ArrayCoder } from './array';

type Model = AdaptyProduct;
type Serializable = Record<string, any>;

export class AdaptyProductCoder extends Coder<Model, Serializable> {
  protected properties: Properties<Model, Serializable> = {
    // version: {
    //   key: 'version',
    //   required: true,
    //   type: 'string',
    // },
    vendorProductId: {
      key: 'vendor_product_id',
      required: true,
      type: 'string',
    },
    localizedDescription: {
      key: 'localized_description',
      required: true,
      type: 'string',
    },
    localizedTitle: {
      key: 'localized_title',
      required: true,
      type: 'string',
    },
    price: {
      key: 'price',
      required: true,
      type: 'number',
    },
    currencyCode: {
      key: 'currency_code',
      required: false,
      type: 'string',
    },
    currencySymbol: {
      key: 'currency_symbol',
      required: false,
      type: 'string',
    },
    subscriptionPeriod: {
      key: 'subscription_period',
      required: false,
      type: 'object',
      converter: new AdaptyProductSubscriptionPeriodCoder(),
    },
    introductoryDiscount: {
      key: 'introductory_discount',
      required: false,
      type: 'object',
      converter: new AdaptyProductDiscountCoder(),
    },
    localizedPrice: {
      key: 'localized_price',
      required: false,
      type: 'string',
    },
    localizedSubscriptionPeriod: {
      key: 'localized_subscription_period',
      required: false,
      type: 'string',
    },
    variationId: {
      key: 'variation_id',
      required: false,
      type: 'string',
    },
    paywallABTestName: {
      key: 'paywall_ab_test_name',
      required: false,
      type: 'string',
    },
    paywallName: {
      key: 'paywall_name',
      required: false,
      type: 'string',
    },
    ios: {
      regionCode: {
        key: 'region_code',
        required: false,
        type: 'string',
      },
      promotionalOfferId: {
        key: 'promotional_offer_id',
        required: false,
        type: 'string',
      },
      promotionalOfferEligibility: {
        key: 'promotional_offer_eligibility',
        required: false,
        type: 'boolean',
      },
      isFamilyShareable: {
        key: 'is_family_shareable',
        required: false,
        type: 'boolean',
      },
      subscriptionGroupIdentifier: {
        key: 'subscription_group_identifier',
        required: false,
        type: 'string',
      },
      discounts: {
        key: 'discounts',
        required: false,
        type: 'array',
        converter: new ArrayCoder(AdaptyProductDiscountCoder as any),
      },
    },
    android: {
      freeTrialPeriod: {
        key: 'free_trial_period',
        required: false,
        type: 'object',
        converter: new AdaptyProductSubscriptionPeriodCoder(),
      },
      localizedFreeTrialPeriod: {
        key: 'localized_free_trial_period',
        required: false,
        type: 'string',
      },
    },
  };
}
